### PR TITLE
Fix "Hide forks/replies until the user expands them"

### PR DESCRIPTION
**Pull request by Powersource**
## What's the problem you solved?

Fixes https://github.com/fraction/oasis/issues/220

## What solution are you recommending?

This one :D Or well this but a bit cleaned up, in this first commit I've included my failed attempts for posterity. You could also argue that this is pretty hacky and should be rewritten lol, and I'd partially agree but tbh I don't know if the other solutions  would be much cleaner.

Good thread for testing (many forks) http://localhost:3000/thread/%25YiBecWPnG9iaftO6xRbzsUYxjYOBuZCBI5IEVnH%2FUc4%3D.sha256

## Todo

* [x] Remove old code and comments
* [x] Make sure all views are rendered with this, I think atm it might just be the specific thread view but idk
* [x] Add `<summary>`ies, using e.g. the first part of the first reply in the fork.
* [x] Make it a bit prettier/more understandable. E.g. right now the `Details` text has 0 left margin, and I'd also like a border on the left of the `<details>`. And since the posts are rendered inside the Details we could move the fork-posts' left margin to the Details.
* [x] If a message in a fork is linked to, the fork should be pre-expanded

### CODE REVIEW

**Comment by christianbundy** [comment]
:tada: Thanks for this patch! I've got a friend visiting from out of town today so I might be slow to check it out, but I'm really excited to have this functionality.

**Comment by Powersource** [comment]
Yeah no pressure, I've got a bunch of changes needed before I'm happy with this :)

**Comment by Powersource** [comment]
Ran into a fun thing where the typescript checker tricked me into removing `.flat()` in one spot because I thought it had figured out it wasn't needed, but that's apparently because we're using an ["old" version of ts](https://stackoverflow.com/questions/53556409/typescript-flatmap-flat-flatten-doesnt-exist-on-type-any). Anyway fixed that by just using `lodash.flatten` instead.

**Comment by christianbundy** [comment]
FWIW I don't think Node 10 has `Array.prototype.flat()` either way.

**Comment by Powersource** [comment]
Yeah I think the issue was pretty much that, I forgot that we're supporting LTS.

**Comment by Powersource** [comment]
a screenshot, now that there's some styling so it's a bit more understandable what's happening

image

**Comment by Powersource** [comment]
image

**Comment by christianbundy** [comment]
Just took this for a spin and it works really great. The thread view and Summaries view are the only ones with nesting, so we're good to go from that angle. The only big todo is making sure that we auto-expand all parent `<details>` when linking to a message.

**Comment by christianbundy** [comment]
Please let me know when you'd like me to take a look at this again! Since it's an open PR I get a notification each time you push, but I'm not sure when it's ready for review.

**Comment by Powersource** [comment]
Yep should be ready for review now!

**Comment by Powersource** [comment]
Wow that was quick! Nice to have this in :)

Was wondering if you had opinions on the preprocessing step/`lodash.set(...meta...)` being outside of models.js?

**Comment by christianbundy** [comment]
Awesome! Thanks a bunch for this patch. I'm going to add you as a maintainer. Please feel free to check out the [development process](https://github.com/fraction/oasis/blob/master/docs/contributing.md#24-development-process), and either create an issue or reach out if you're not sure about something.

Welcome aboard. :train: :train:

**Comment by Powersource** [comment]
Thanks! And tbh I'm not gonna be able to remember that whole process document but I'll try :sweat_smile: Just tell me if something's off.

Also bumping my q since I'm guessing you missed it when we commented at the same time :)  
> Was wondering if you had opinions on the preprocessing step/lodash.set(...meta...) being outside of models.js?

**Comment by christianbundy** [comment]
TL;DR: Please merge any branch that passes existing tests and solves a problem, except when it causes another problem.

> Was wondering if you had opinions on the preprocessing step/lodash.set(...meta...) being outside of models.js?

As a maintainer I don't have opinions on implementation details, but as a contributor I think I'd probably like to see that sort of thing in the models. I've been thinking that I'd like to write a patch to refactor the models so that they return nested arrays of threads rather than `depth`, that way we don't need to handle this in the views, but I haven't had the time or energy to write that patch yet. If you have thoughts, I'd be super interested to hear them!

**Comment by christianbundy** [comment]
And yep, I totally missed your question, thanks for pointing it out. :100:
